contracts/liquidity_hub/pool-network/terraswap_pair/src/helpers.rs

```diff
@@ -5,44 +5,345 @@ use cosmwasm_schema::cw_serde;
+            (ann.checked_sub(Decimal256::one())?
+                .checked_mul(current_d)?
+                .checked_add(n_coins.checked_add(Decimal256::one())?.checked_mul(new_d)?))?,
+        )?;
+
+        if current_d >= old_d {
+            if current_d.checked_sub(old_d)? <= Decimal256::decimal_with_precision(1u8, precision)?
+            {
+                // success
+                return Ok(current_d);
+            }
+        } else if old_d.checked_sub(current_d)?
+            <= Decimal256::decimal_with_precision(1u8, precision)?
+        {
+            // success
+            return Ok(current_d);
+        }
+    }
+
+    // completed iterations
+    // but we never approximated correctly
+    Err(ContractError::ConvergeError {})
+}
+
+/// Determines the direction of `offer_pool` -> `ask_pool`.
+///
+/// In a `ReverseSimulate`, we subtract the `offer_pool` from `offer_amount` to get the pool sum.
+///
+/// In a `Simulate`, we add the two.
+pub enum StableSwapDirection {
+    Simulate,
+    ReverseSimulate,
+}
+
+/// Calculates the new pool amount given the current pools and swap size.
+pub fn calculate_stableswap_y(
+    offer_pool: Decimal256,
+    ask_pool: Decimal256,
+    offer_amount: Decimal256,
+    amp: &u64,
+    offer_precision: u8,
+    ask_precision: u8,
+    direction: StableSwapDirection,
+) -> Result<Uint128, ContractError> {
+    let ann = Uint256::from_u128((*amp).into()).checked_mul(N_COINS)?;
+
+    let d = calculate_stableswap_d(offer_pool, ask_pool, amp, ask_precision)?
+        .to_uint256_with_precision(u32::from(ask_precision))?;
+
+    let pool_sum = match direction {
+        StableSwapDirection::Simulate => offer_pool.checked_add(offer_amount)?,
+        StableSwapDirection::ReverseSimulate => ask_pool.checked_sub(offer_amount)?,
+    }
+    .to_uint256_with_precision(u32::from(offer_precision.max(ask_precision)))?;
+    let c = d
+        .checked_multiply_ratio(d, pool_sum.checked_mul(N_COINS)?)?
+        .checked_multiply_ratio(d, ann.checked_mul(N_COINS)?)?;
+
+    let b = pool_sum.checked_add(d.checked_div(ann)?)?;
+
+    // attempt to converge solution using Newton-Raphson method
+    let mut y = d;
+    for _ in 0..NEWTON_ITERATIONS {
+        let previous_y = y;
+        // y = (y^2 + c) / (2y + b - d)
+        y = y
+            .checked_mul(y)?
+            .checked_add(c)?
+            .checked_div(y.checked_add(y)?.checked_add(b)?.checked_sub(d)?)?;
+
+        if y >= previous_y {
+            if y.checked_sub(previous_y)? <= Uint256::one() {
+                return y
+                    .try_into()
+                    .map_err(|_| ContractError::SwapOverflowError {});
+            }
+        } else if y < previous_y && previous_y.checked_sub(y)? <= Uint256::one() {
+            return y
+                .try_into()
+                .map_err(|_| ContractError::SwapOverflowError {});
+        }
+    }
+
+    Err(ContractError::ConvergeError {})
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
What about moving these tests to `tests/testing.rs`?

Reply:
Yep, I will move it to a relevant file ✔️